src/models/tree/build_tree.h

```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
   return histogram_node == node_id;
 }
 
-__host__ __device__ inline auto CalculateLeafValue(double G, double H, double alpha) -> double
+namespace detail {
+__host__ __device__ inline auto CalculateLeafGain(const GPair& sum,
+                                                  double alpha_reg,
+                                                  double l1_regularization) -> double
 {
-  return -G / (H + alpha);
+  // l1 threshold
+  double z = 0.0;
+  if (sum.grad > l1_regularization) {
+    z = sum.grad - l1_regularization;
+  } else if (sum.grad < -l1_regularization) {
+    z = sum.grad + l1_regularization;
+  }
+  // Note: we could use the cuda __ddiv_rn intrinsic here if this becomes a bottleneck
+  return (z * z) / (sum.hess + alpha_reg);
+}
+}  // namespace detail
+
+__host__ __device__ inline auto CalculateGain(const GPair& left_sum,
+                                              const GPair& right_sum,
+                                              const GPair& parent_sum,
+                                              double alpha,
+                                              double l1_regularization,
+                                              double gamma) -> double
+{
+  double const reg = std::max(1e-5, alpha);
```

Comment:
Why not keep using `eps`?

Reply:
I had a compiler error where you cannot take a reference to static const variable in device code - std::max takes a reference. Its fixed by doing `double const reg = std::max(double(eps), alpha);` to force a copy.